libcbtest/callback.c

```diff
@@ -0,0 +1,55 @@
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <unistd.h>
+
+typedef int (*callback)(const char *, int);
+
+struct node {
+    const void *fp;
+    char *s;
+    struct node* next;
+};
+
+pthread_cond_t cv;
+pthread_mutex_t mp;
+static struct node *head = NULL;
+
+void* startThread(void *arg) {
+    printf("Start thread in lib\n");
+    for (;;) {
+        pthread_mutex_lock(&mp);
+        pthread_cond_wait(&cv, &mp);
+        if (head != NULL) {
+            struct node *n = head;
+            head = head->next;
+
+            printf("FROM C: %s\n", n->s);
+            int size = ((callback)(n->fp))(n->s, strlen(n->s));
+            printf("Back from callback: %d\n", size);
+
+            free(n->s);
+            free(n);
+        }
+        pthread_mutex_unlock(&mp);
```

Comment:
Do we need manipulating nodes in order to reproduce the issue? Or would just using a mutex be fine?

Reply:
The crash happens after the Go callback returns and the C code access local variables because the frame pointer is incorrect. It was strange when the thread start function is compiled by cgo and is part of the main Go binary, it did not crash. I checked that FP was still overwritten but did not crash. FP was replaced by LR since the crosscall2 frame overlaps by one machine word, so it may just be a coincidence. Access a local variable that points to heap memory is more likely to trigger the crash.

Reply:
If we can get the minimized test to reproduce the issue, this would be the best. I'm still not sure the algorithm for the nodes and/or printf outputs. If these are likely needed, please leave a comment to explain why we need these code.

Reply:
Turns out just checking a local variable against a known value is good enough. Just need to make sure to disable optimization so the C compiler does not optimize out the local variable.